proposals/000-streamer-api.md

```diff
@@ -0,0 +1,520 @@
+      * parse exploded archives or in archiving formats unknown to the parser, e.g. RAR
+    * If the file is not an archive, it will be reachable at the HREF `/publication.<file.format.fileExtension>`, e.g. with a PDF.
+  * `fallbackTitle: String = file.name`
+    * The `Publication`'s `title` is mandatory, but some formats might not have a way of declaring a title (e.g. CBZ). In which case, `fallbackTitle` will be used.
+    * The default implementation uses the filename as the fallback title.
+  * `warnings: WarningLogger? = null`
+    * Logger used to broadcast non-fatal parsing warnings.
+    * Can be used to report publication authoring mistakes, to warn users of potential rendering issues or help authors debug their publications.
+
+
+### `Streamer` Class
+
+Opens a `Publication` using a list of parsers.
+
+#### Constructor
+
+* `Streamer(/* see parameters below */)`
+  * `parsers: List<PublicationParser> = []`
+    * Parsers used to open a publication, in addition to the default parsers.
+    * The provided parsers take precedence over the default parsers.
+    * This can be used to provide custom parsers, or a different configuration for default parsers.
+  * `ignoresDefaultParsers: Boolean = false`
+    * When `true`, only parsers provided in `parsers` will be used.
+    * Can be used if you want to support only a subset of Readium's parsers.
+  * `httpClient: HTTPClient? = default`
+    * HTTP client used to perform any HTTP requests.
+    * The default implementation uses native APIs when available.
+  * `openArchive: Archive.Factory? = default`
+    * Opens an archive (e.g. ZIP, RAR), optionally protected by credentials.
+    * The default implementation uses native APIs when available.
+  * `openXML: XMLDocument.Factory? = default`
+    * Parses an XML document into a DOM tree.
+    * The default implementation uses native APIs when available.
+  * `openPDF: PDFDocument.Factory? = default`
+    * Parses a PDF document, optionally protected by password.
+    * The default implementation uses native APIs when available.
+  * `onCreateManifest: (File, Manifest) -> Manifest = { f, m -> m }`
+    * Called before creating the `Publication`, to modify the parsed `Manifest` if desired.
+  * `onCreateFetcher: (File, Manifest, Fetcher) -> Fetcher = { f, m, fetcher -> fetcher }`
+    * Called before creating the `Publication`, to modify its root fetcher.
+  * `onCreateServices: (File, Manifest, Publication.ServicesBuilder) -> Void = { f, m, sb -> }`
+    * Called before creating the `Publication`, to modify its list of service factories.
+
+The specification of `HTTPClient`, `Archive`, `XMLDocument` and `PDFDocument` is out of scope for this proposal.
+
+#### Methods
+
+* `open(file: File, fallbackTitle: String = file.name, warnings: WarningLogger? = null) -> Future<Publication?>`
+  * Parses a `Publication` from the given `file`.
+  * Returns `null` if the file was not recognized by any parser, or a `Streamer.Error` in case of failure.
```

Comment:
After experimenting in the testapp, this is not convenient. I think `open` may return a custom `Streamer.ParsingFailed` when no parser is able to parse a publication. So, all errors can be handled in the same way.

Reply:
Sounds reasonable, but I would introduce a separated error for that, such as `UnsupportedFormat`. Typically the app will want to show these two errors separately: "Corrupted/unreadable file" and "Unsupported publication format". The details of `ParsingFailed` are kept for logging but probably not shown to the user.